src/DominantJuice.sol

```diff
@@ -0,0 +1,396 @@
+import {IJBSingleTokenPaymentTerminal} from
+    "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBSingleTokenPaymentTerminal.sol";
+import {IJBSingleTokenPaymentTerminalStore3_1_1} from
+    "@jbx-protocol/juice-contracts-v3/contracts/interfaces/IJBSingleTokenPaymentTerminalStore3_1_1.sol";
+import {JBDidPayData3_1_1} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBDidPayData3_1_1.sol";
+import {JBDidRedeemData3_1_1} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBDidRedeemData3_1_1.sol";
+import {JBFundingCycle} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBFundingCycle.sol";
+import {JBPayDelegateAllocation3_1_1} from
+    "@jbx-protocol/juice-contracts-v3/contracts/structs/JBPayDelegateAllocation3_1_1.sol";
+import {JBPayParamsData} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBPayParamsData.sol";
+//import {JBDidRedeemData} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBDidRedeemData.sol";
+import {JBRedeemParamsData} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBRedeemParamsData.sol";
+import {JBRedemptionDelegateAllocation3_1_1} from
+    "@jbx-protocol/juice-contracts-v3/contracts/structs/JBRedemptionDelegateAllocation3_1_1.sol";
+import {JBTokenAmount} from "@jbx-protocol/juice-contracts-v3/contracts/structs/JBTokenAmount.sol";
+import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
+import {UD60x18, ud, add, pow, powu, div, mul, wrap, unwrap} from "@prb/math/UD60x18.sol";
+
+/**
+ * @title Dominant Juice
+ * @author Armand Daigle
+ * @notice This contract is a JB Data Source, JB Pay Delegate, JB Redemption Delegate, and a Refund Bonus Escrow
+ * implementation. Campaigns, referred to as cycles from here on, that use this implementation as a Data Source
+ * will incorporate Alex Tabarrok's dominant assurance mechanism, which incentivize early pledgers with a campaign
+ * creator-provided refund bonus that can be withdrawn from this contract after a failed campaign.
+ * @dev This contract uses the JB Data Source, Pay Delegate, and Pay Redemption interfaces and extends the
+ * functionality of a Juicebox project. There is no constructor since each campaign will deploy a clone of this
+ * contract. The initialize() function acts as the constructor. This contract is written to only be used for one cycle.
+ * If projects want dominant assurance on more cycles, refactoring is required.
+ */
+contract DominantJuice is IJBFundingCycleDataSource3_1_1, IJBPayDelegate3_1_1, IJBRedemptionDelegate3_1_1, Ownable {
+    /// Juicebox Contracts
+
+    IJBController3_1 public controller;
+    IJBDirectory public directory; // directory of terminals and controllers for projects.
+    IJBFundAccessConstraintsStore public fundAccessConstraintsStore;
+    IJBSingleTokenPaymentTerminalStore3_1_1 public paymentTerminalStore;
+    IJBSingleTokenPaymentTerminal public paymentTerminal;
+
+    /// Juicebox Project State Variables
+    uint256 public projectId;
```

Comment:
also, should all of these really be public? Making them public just clutters the "public API" of the contract. Think about what you want folks to see about this contract and make those things public or create custom public view functions to return that info.

Reply:
> there are a lot of top-level variables that would be good within one or more structs

Certainly, I can put in structs. May I ask what advantages struct use offers here, or what disadvantages it avoids?


> also, should all of these really be public? Making them public just clutters the "public API" of the contract. Think about what you want folks to see about this contract and make those things public or create custom public view functions to return that info.

I can definitely do a pass and make as many of them private as I can. And besides decluttering the API, are there any other advantages to make all the variables you can private?

Reply:
The advantages of using a struct are just maintainability. Having a bunch of loose unrelated fields makes it difficult to understand what they are used for. Building up structures allows the reader to see what fields go together and how they are used. It can be less important if you break up the functionality into multiple classes, but since all of the code is in one contract, it would be better to break up the functionality into logical groupings if possible.

As for public vs. private, all blockchain data is public to anyone who wants to inspect it. However, if it's public then other contracts on-chain can access it. If it's private, then other contracts can't access it. Making data private can help prevent some forms of on-chain gaming.